### DISCUSSION TITLE
Compiling fdpp produces 'Have 66 undefined symbols'

### DISCUSSION THREAD

**Discussion by JohnCC330** [dosemu2/dosemu2#1648] [General]
On a virtual machine with Slackware64-current installed, I tried to compiling fdpp by just typing 'make'. After a while, I get the message  'Have 66 undefined symbols' and compilation stopped. I went back as far as the backscroll permitted and did not see any error messages. So I suspect this error appears on linking.

The last command is clang++ -olibfdpp.so ... 

A bit of web searching didn't help. Any suggestions?

clang version 13.0.1

**Comment by stsp** [comment]
> A bit of web searching didn't help. Any suggestions?

Unless you tell what exactly doesn't
work, there is very little that can be
done for a help.
After `make` you should do
`sudo make install`

**Reply by JohnCC330** [reply]
Not to brag, but I've maybe compiled/installed a few thousand programs in Linux ;)
If I get a message `Have 66 undefined symbols` I tend to believe the compilation (linking) failed and do not install. Is `fdpp` and exception? Will these references be resolved later?

**Reply by stsp** [reply]
Any binary (executable or library)
have undefined symbols if it is
linked to a shared library.
It is what dynamic linker (ld.so)
for: to resolve all undefined symbols.

I realize you ask in a forum, not
in an issue ticket, but you present
it as a problem, not as a question.
Any problem should be accomplished
with the exact info on what doesn't
work.

**Reply by stsp** [reply]
> Not to brag, but I've maybe compiled/installed a few thousand programs in Linux ;)

Then you know how exactly
`make` reports build errors.
`make: *** <Error msg here>.  Stop.`

**Reply by JohnCC330** [reply]
My apologies for the delay.

This is the final part of `make`:

```
make[2]: Leaving directory '/root/Downloads/fdpp-master/fdpp/parsers'
parsers/thunk_gen <thunk_calls.tmp >thunk_calls.h || (rm -f thunk_calls.h ; false)
/root/Downloads/fdpp-master/fdpp/parsers/gen_thunks.sh thunk_asms.tmp /root/Downloads/fdpp-master/fdpp/parsers >thunk_a
        (rm -f thunk_asms.h ; false)
clang++ -std=c++11 -c -fno-threadsafe-statics -fpic -iquote /root/Downloads/fdpp-master/fdpp/../hdr -DFDPP -DDEBUG -DWI
clang++ -std=c++11 -c -fno-threadsafe-statics -fpic -iquote /root/Downloads/fdpp-master/fdpp/../hdr -DFDPP -DDEBUG -DWI
clang++ -std=c++11 -c -fno-threadsafe-statics -fpic -iquote /root/Downloads/fdpp-master/fdpp/../hdr -DFDPP -DDEBUG -DWI
clang++ -std=c++11 -c -fno-threadsafe-statics -fpic -iquote /root/Downloads/fdpp-master/fdpp/../hdr -DFDPP -DDEBUG -DWI
clang++ -std=c++11 -c -fno-threadsafe-statics -fpic -iquote /root/Downloads/fdpp-master/fdpp/../hdr -DFDPP -DDEBUG -DWI
clang++ -std=c++11 -c -fno-threadsafe-statics -fpic -iquote /root/Downloads/fdpp-master/fdpp/../hdr -DFDPP -DDEBUG -DWI
clang++ -o libfdpp.so blockio.o break.o chario.o dosfns.o dsk.o error.o fatdir.o fatfs.o fattab.o fcbfns.o inthndlr.o i
Have 66 undefined symbols
sed \
        -e 's!@PREFIX[@]!/usr/local!g' \
        -e 's!@INCLUDEDIR[@]!/usr/local/include!g' \
        -e 's!@LIBDIR[@]!/usr/local/lib!g' \
        -e 's!@VERSION[@]!1.6!g' \
        /root/Downloads/fdpp-master/fdpp/fdpp.pc.in >fdpp.pc
make[1]: Leaving directory '/root/Downloads/fdpp-master/fdpp'
```
I've detected that downloading the code as 'zip' file seems to have caused this. The `zip`does not contain a complete git repo.
A cloned copy does compile correctly.

I could not find a way to install in the 'slackware' - way, i.e. in `/usr` and `/usr/lib64`.

Anyway, dosemu seems to be running, though probably I'll need to edit the configuration somewhat.

Cheers.
[complete log: log.gz](https://github.com/dosemu2/dosemu2/files/8826210/log.gz)

**Reply by stsp** [reply]
> A cloned copy does compile correctly.

Not sure why you write this.
This is a false statement, and
you know that.
There is no difference wrt that
line, whether you compile from
git or not.

I guess I should stop replying
to this until you either present
the problem, or tell that everything
works.
